sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -475,6 +476,58 @@ object RemoveRedundantAliases extends Rule[LogicalPlan] {
   def apply(plan: LogicalPlan): LogicalPlan = removeRedundantAliases(plan, AttributeSet.empty)
 }
 
+/**
+ * Remove redundant aggregates from a query plan. A redundant aggregate is an aggregate whose
+ * only goal is to keep distinct values, while its parent aggregate would ignore duplicate values.
+ */
+object RemoveRedundantAggregates extends Rule[LogicalPlan] with AliasHelper {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transformUp {
+    case upper @ Aggregate(_, _, lower: Aggregate) if lowerIsRedundant(upper, lower) =>
+      val aliasMap = getAliasMap(lower)
+      upper.copy(
+        child = lower.child,
+        groupingExpressions = upper.groupingExpressions.map(replaceAlias(_, aliasMap)),
+        aggregateExpressions = upper.aggregateExpressions.map(
+          replaceAliasButKeepOuter(_, aliasMap))
+      )
+  }
+
+  private def lowerIsRedundant(upper: Aggregate, lower: Aggregate): Boolean = {
+    val isDeterministic = upper.aggregateExpressions.forall(_.deterministic) &&
+      lower.aggregateExpressions.forall(_.deterministic)
```

Comment:
This check looks strict, so could you make it looser? e.g., how about the case where `lower.aggregateExpressions` has non-determinisitc exprs, but a upper node does not reference them.

Reply:
I was thinking about some stateful UDF. As a dummy example - one that counts the number of times it has been called and returns that count. Even if it is not referenced, evaluating it changes future results. 
Or perhaps I'm overthinking it and we can ignore it? 
Taking it a step further - perhaps only upper has to be deterministic?

Reply:
> As a dummy example - one that counts the number of times it has been called and returns that count. Even if it is not referenced, evaluating it changes future results.

I think `ColumnPruning` removes exprs not referenced by a upper node even in the case above, so we can ignore it.
```
// v3.0.1
scala> sql("select a, b from (select a, b, rand() from t group by a, b) group by a, b").explain(true)
== Analyzed Logical Plan ==
a: bigint, b: bigint
Aggregate [a#0L, b#1L], [a#0L, b#1L]
+- SubqueryAlias __auto_generated_subquery_name
   +- Aggregate [a#0L, b#1L], [a#0L, b#1L, rand(-8085205063662585668) AS rand(-8085205063662585668)#31]
      +- SubqueryAlias spark_catalog.default.t
         +- Relation[a#0L,b#1L] parquet

== Optimized Logical Plan ==
Aggregate [a#0L, b#1L], [a#0L, b#1L]
+- Aggregate [a#0L, b#1L], [a#0L, b#1L]
   +- Relation[a#0L,b#1L] parquet
```
```
```

Reply:
> Taking it a step further - perhaps only upper has to be deterministic?

How about the case below? Seems like we can remove the lower agg?
```
scala> sql("select a, b, rand() from (select a, b from t group by a, b) group by a, b").explain(true)
== Optimized Logical Plan ==
Aggregate [a#0L, b#1L], [a#0L, b#1L, rand(3153957949044950464) AS rand(3153957949044950464)#2]
+- Aggregate [a#0L, b#1L], [a#0L, b#1L]
   +- Relation[a#0L,b#1L] parquet
```

Reply:
Probably, we can re-use `CollapseProject#haveCommonNonDeterministicOutput`?  Seems like the similar pre-condition to collapse aggregates, I think.

Reply:
I reused the `CollapseProject#haveCommonNonDeterministicOutput`, but now I have hard time figuring out a case where it would have any effect, because the grouping expression can't be non-deterministic (`PullOutNondeterministic` analyzer rule) and we have condition `upperReferencesOnlyGrouping`.

Reply:
Ah, yes, you're right. I've checked the related code again and I think we don't need to check if exprs are deterministic or not because of `PullOutNondeterministic`. Seems like it is okay just to check if grouping exprs in a upper node is a subset of grouping exprs in a lower node.

Reply:
I removed the determinicity check. For that I had to move the `PullOutNondeterministic` analyzer rule around a bit. Does it look okay?